File: qc-domain/src/main/java/com/eager/questioncloud/domain/coupon/repository/UserCouponRepository.java
package com.eager.questioncloud.domain.coupon.repository;

import com.eager.questioncloud.domain.coupon.dto.AvailableUserCoupon;
import com.eager.questioncloud.domain.coupon.model.UserCoupon;
import java.util.List;

public interface UserCouponRepository {
    UserCoupon getUserCoupon(Long userCouponId, Long userId);

    UserCoupon getUserCoupon(Long userCouponId);

    Boolean isRegistered(Long userId, Long couponId);

    UserCoupon save(UserCoupon userCoupon);

    List<AvailableUserCoupon> getAvailableUserCoupons(Long userId);
}

File: qc-domain/qc-coupon/src/main/kotlin/com/eager/questioncloud/coupon/infrastructure/repository/UserCouponRepository.kt
package com.eager.questioncloud.coupon.infrastructure.repository

import com.eager.questioncloud.coupon.domain.UserCoupon
import com.eager.questioncloud.coupon.dto.AvailableUserCoupon

interface UserCouponRepository {
    fun getUserCoupon(userCouponId: Long, userId: Long): UserCoupon

    fun getUserCoupon(userCouponId: Long): UserCoupon

    fun isRegistered(userId: Long, couponId: Long): Boolean

    fun save(userCoupon: UserCoupon): UserCoupon

    fun getAvailableUserCoupons(userId: Long): List<AvailableUserCoupon>

    fun use(userCouponId: Long): Boolean

    fun deleteAllInBatch()
}
